File: coding-challange/codewars/6kyu/2020-04-15~2020-06-01/find-the-missing-term-in-an-arithmetic-progression/find-the-missing-term-in-an-arithmetic-progression.py
# find-the-missing-term-in-an-arithmetic-progression
# https://www.codewars.com/kata/52de553ebb55d1fca3000371/

from unittest import TestCase


def find_missing(sequence):
    [a, b] = [sequence[1] - sequence[0], sequence[2] - sequence[1]]
    rule = a if abs(a) < abs(b) else b

    for i in range(len(sequence) - 1):
        if sequence[i] + rule != sequence[i + 1]:
            return sequence[i] + rule


TestCase().assertEqual(find_missing([1, 2, 3, 4, 6, 7, 8, 9]), 5)
TestCase().assertEqual(find_missing([1, 3, 4, 5, 6, 7, 8, 9]), 2)

File: coding-challange/codewars/6kyu/2020-04-15~2020-06-01/find-the-missing-term-in-an-arithmetic-progression/find-the-missing-term-in-an-arithmetic-progression.js
// Find the missing term in an Arithmetic Progression
// https://www.codewars.com/kata/52de553ebb55d1fca3000371/

const findMissing = (list) => {
  const [a, b] = [list[1] - list[0], list[2] - list[1]];
  const rule = Math.abs(a) < Math.abs(b) ? a : b;
  
  return list.reduce((acc, cur, i, arr) => {
    if (acc + rule !== cur) {
      arr.splice(0);
      return acc + rule;
    }
    return cur;
  });
};


console.log(findMissing([1, 3, 4]), 2);
console.log(findMissing([-23, -27, -29, -31, -33]), -25);
